mpcite/doi_builder.py

```diff
@@ -0,0 +1,487 @@
+from urllib3.exceptions import HTTPError
+from datetime import datetime
+from tqdm import tqdm
+from maggma.stores import Store, MongoStore
+
+
+class DoiBuilder(Builder):
+    """
+    A builder to combine information from the Materials Project database(MP) and Osti Explorer(OSTI) to produce a *dois*
+    collections
+
+    *dois* collection is similar to a "cache" where it maps the mp ids that are in MP to the DOIs that are in the OSTI
+
+    This builder will find all materials that are in MP but are not in the  in the *dois* collection, call this list of
+    uncited_materials
+
+    It will submit the uncited_materials to OSTI and constantly check whether OSTI has updated its database
+
+    if it finds one of the uncited_materials in OSTI, it will update the *dois* collection
+
+    From a high level, this Builder works as follows:
+    1. get_items -> get all the records that require registration / update
+    2. process_item -> compute the update to one particular data from get_items
+    3. update_targets -> batch update DOI collection and update remote services
+
+    Please note that the default run() function ties the above three functions together
+
+    """
+
+    def __init__(self,
+                 materials_store: Store,
+                 robocrys_store: Store,
+                 doi_store: Store,
+                 osti: OSTIModel,
+                 elsevier: ConnectionModel,
+                 max_doi_requests=1000,
+                 sync=True,
+                 **kwargs):
+        """
+         connection with materials database
+            1. establish connection with materials collection (Guaranteed online)
+            2. establish connection with doi collection (online or local)
+            3. establish connection with robocrys (Guaranteed online)
+
+        establish connection with ELink to submit info
+
+        establish connection with osti explorer to get bibtex
+
+
+        :param adapter: OstiMongoAdapter that keeps track of materials, doi, and other related stores
+        :param osti: osti connection specification
+        :param max_doi_requests: the maximum send size that E-Link allows
+        :param sync: boolean representing whether to sync DOI collection with Elink and Explorer in get_items
+        :param kwargs: other keywords fed into Builder(will be documented as development goes on)
+        """
+        super().__init__(sources=[materials_store, robocrys_store],
+                         targets=[doi_store],
+                         **kwargs)
+        # set connections
+        self.materials_store = materials_store
+        self.robocrys_store = robocrys_store
+        self.doi_store = doi_store
+        self.osti = osti
+        self.elsevier = elsevier
+        self.elink_adapter = ELinkAdapter(osti.elink)
+        self.explorer_adapter = ExplorerAdapter(osti.explorer)
+        self.elsevier_adapter = ElviserAdapter(self.elsevier)
```

Comment:
No point in the "OSTI" container if you're just going to unravel it here. Make these the init args: `elink_adapter`, etc.